File: benchmark/haskell/BenchmarkExp.hs
module Main where

import Control.Monad
import qualified Data.List
import qualified MergesortHaskellNTRStack
import qualified MergesortHaskellNTRStack_
import qualified MergesortHaskellTRStack
import qualified MergesortHaskellTRStack_
import qualified MergesortHaskellStdlib
import qualified MergesortCoqCbn
import qualified MergesortCoqCbnAcc
import qualified MergesortCoqCbv
import qualified MergesortCoqCbvAcc
import qualified GHC.Stats

import Benchlib

main :: IO ()
main = do
  statsEnabled <- GHC.Stats.getRTSStatsEnabled
  unless statsEnabled (error "+RTS -T required.")
  benchmark
    "haskell-partial-random" (map (\i -> 128 * 2 ^ i) [0..14]) id
    [("Data.List.sort (old)", \n xs -> sorted (take (n `div` 4) (MergesortHaskellStdlib.oldSort (<=) xs))   `seq` return ()),
     ("Data.List.sort (new)", \n xs -> sorted (take (n `div` 4) (MergesortHaskellStdlib.newSort (<=) xs))   `seq` return ()),
     ("NTRStack.sort3",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack.sort3 (<=) xs))   `seq` return ()),
     ("NTRStack.sortN",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack.sortN (<=) xs))   `seq` return ()),
     ("NTRStack.sort3N",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack.sort3N (<=) xs))  `seq` return ()),
     ("NTRStack_.sort3",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack_.sort3 (<=) xs))  `seq` return ()),
     ("NTRStack_.sortN",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack_.sortN (<=) xs))  `seq` return ()),
     ("NTRStack_.sort3N",     \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack_.sort3N (<=) xs)) `seq` return ()),
     ("TRStack.sort3",        \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack.sort3 (<=) xs))    `seq` return ()),
     ("TRStack.sortN",        \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack.sortN (<=) xs))    `seq` return ()),
     ("TRStack.sort3N",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack.sort3N (<=) xs))   `seq` return ()),
     ("TRStack_.sort3",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack_.sort3 (<=) xs))   `seq` return ()),
     ("TRStack_.sortN",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack_.sortN (<=) xs))   `seq` return ()),
     ("TRStack_.sort3N",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack_.sort3N (<=) xs))  `seq` return ()),
     ("CBNAcc.sort2",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbnAcc.sort2 (<=) xs))         `seq` return ()),
     ("CBNAcc.sort3",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbnAcc.sort3 (<=) xs))         `seq` return ()),
     ("CBNAcc.sortN",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbnAcc.sortN (<=) xs))         `seq` return ()),
     ("CBVAcc.sort2",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbvAcc.sort2 (<=) xs))         `seq` return ()),
     ("CBVAcc.sort3",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbvAcc.sort3 (<=) xs))         `seq` return ()),
     ("CBVAcc.sortN",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbvAcc.sortN (<=) xs))         `seq` return ())]
  benchmark
    "haskell-partial-smooth" (map (\i -> 128 * 2 ^ i) [0..14]) (sort_blocks 50)
    [("Data.List.sort (old)", \n xs -> sorted (take (n `div` 4) (MergesortHaskellStdlib.oldSort (<=) xs))   `seq` return ()),
     ("Data.List.sort (new)", \n xs -> sorted (take (n `div` 4) (MergesortHaskellStdlib.newSort (<=) xs))   `seq` return ()),
     ("NTRStack.sort3",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack.sort3 (<=) xs))   `seq` return ()),
     ("NTRStack.sortN",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack.sortN (<=) xs))   `seq` return ()),
     ("NTRStack.sort3N",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack.sort3N (<=) xs))  `seq` return ()),
     ("NTRStack_.sort3",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack_.sort3 (<=) xs))  `seq` return ()),
     ("NTRStack_.sortN",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack_.sortN (<=) xs))  `seq` return ()),
     ("NTRStack_.sort3N",     \n xs -> sorted (take (n `div` 4) (MergesortHaskellNTRStack_.sort3N (<=) xs)) `seq` return ()),
     ("TRStack.sort3",        \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack.sort3 (<=) xs))    `seq` return ()),
     ("TRStack.sortN",        \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack.sortN (<=) xs))    `seq` return ()),
     ("TRStack.sort3N",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack.sort3N (<=) xs))   `seq` return ()),
     ("TRStack_.sort3",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack_.sort3 (<=) xs))   `seq` return ()),
     ("TRStack_.sortN",       \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack_.sortN (<=) xs))   `seq` return ()),
     ("TRStack_.sort3N",      \n xs -> sorted (take (n `div` 4) (MergesortHaskellTRStack_.sort3N (<=) xs))  `seq` return ()),
     ("CBNAcc.sort2",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbnAcc.sort2 (<=) xs))         `seq` return ()),
     ("CBNAcc.sort3",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbnAcc.sort3 (<=) xs))         `seq` return ()),
     ("CBNAcc.sortN",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbnAcc.sortN (<=) xs))         `seq` return ()),
     ("CBVAcc.sort2",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbvAcc.sort2 (<=) xs))         `seq` return ()),
     ("CBVAcc.sort3",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbvAcc.sort3 (<=) xs))         `seq` return ()),
     ("CBVAcc.sortN",         \n xs -> sorted (take (n `div` 4) (MergesortCoqCbvAcc.sortN (<=) xs))         `seq` return ())]
  benchmark
    "haskell-total-random" (map (\i -> 128 * 2 ^ i) [0..14]) id
    [("Data.List.sort (old)", \_ xs -> sorted (MergesortHaskellStdlib.oldSort (<=) xs)   `seq` return ()),
     ("Data.List.sort (new)", \_ xs -> sorted (MergesortHaskellStdlib.newSort (<=) xs)   `seq` return ()),
     ("NTRStack.sort3",       \_ xs -> sorted (MergesortHaskellNTRStack.sort3 (<=) xs)   `seq` return ()),
     ("NTRStack.sortN",       \_ xs -> sorted (MergesortHaskellNTRStack.sortN (<=) xs)   `seq` return ()),
     ("NTRStack.sort3N",      \_ xs -> sorted (MergesortHaskellNTRStack.sort3N (<=) xs)  `seq` return ()),
     ("NTRStack_.sort3",      \_ xs -> sorted (MergesortHaskellNTRStack_.sort3 (<=) xs)  `seq` return ()),
     ("NTRStack_.sortN",      \_ xs -> sorted (MergesortHaskellNTRStack_.sortN (<=) xs)  `seq` return ()),
     ("NTRStack_.sort3N",     \_ xs -> sorted (MergesortHaskellNTRStack_.sort3N (<=) xs) `seq` return ()),
     ("TRStack.sort3",        \_ xs -> sorted (MergesortHaskellTRStack.sort3 (<=) xs)    `seq` return ()),
     ("TRStack.sortN",        \_ xs -> sorted (MergesortHaskellTRStack.sortN (<=) xs)    `seq` return ()),
     ("TRStack.sort3N",       \_ xs -> sorted (MergesortHaskellTRStack.sort3N (<=) xs)   `seq` return ()),
     ("TRStack_.sort3",       \_ xs -> sorted (MergesortHaskellTRStack_.sort3 (<=) xs)   `seq` return ()),
     ("TRStack_.sortN",       \_ xs -> sorted (MergesortHaskellTRStack_.sortN (<=) xs)   `seq` return ()),
     ("TRStack_.sort3N",      \_ xs -> sorted (MergesortHaskellTRStack_.sort3N (<=) xs)  `seq` return ()),
     ("CBNAcc.sort2",         \_ xs -> sorted (MergesortCoqCbnAcc.sort2 (<=) xs)         `seq` return ()),
     ("CBNAcc.sort3",         \_ xs -> sorted (MergesortCoqCbnAcc.sort3 (<=) xs)         `seq` return ()),
     ("CBNAcc.sortN",         \_ xs -> sorted (MergesortCoqCbnAcc.sortN (<=) xs)         `seq` return ()),
     ("CBVAcc.sort2",         \_ xs -> sorted (MergesortCoqCbvAcc.sort2 (<=) xs)         `seq` return ()),
     ("CBVAcc.sort3",         \_ xs -> sorted (MergesortCoqCbvAcc.sort3 (<=) xs)         `seq` return ()),
     ("CBVAcc.sortN",         \_ xs -> sorted (MergesortCoqCbvAcc.sortN (<=) xs)         `seq` return ())]
  benchmark
    "haskell-total-smooth" (map (\i -> 128 * 2 ^ i) [0..14]) (sort_blocks 50)
    [("Data.List.sort (old)", \_ xs -> sorted (MergesortHaskellStdlib.oldSort (<=) xs)   `seq` return ()),
     ("Data.List.sort (new)", \_ xs -> sorted (MergesortHaskellStdlib.newSort (<=) xs)   `seq` return ()),
     ("NTRStack.sort3",       \_ xs -> sorted (MergesortHaskellNTRStack.sort3 (<=) xs)   `seq` return ()),
     ("NTRStack.sortN",       \_ xs -> sorted (MergesortHaskellNTRStack.sortN (<=) xs)   `seq` return ()),
     ("NTRStack.sort3N",      \_ xs -> sorted (MergesortHaskellNTRStack.sort3N (<=) xs)  `seq` return ()),
     ("NTRStack_.sort3",      \_ xs -> sorted (MergesortHaskellNTRStack_.sort3 (<=) xs)  `seq` return ()),
     ("NTRStack_.sortN",      \_ xs -> sorted (MergesortHaskellNTRStack_.sortN (<=) xs)  `seq` return ()),
     ("NTRStack_.sort3N",     \_ xs -> sorted (MergesortHaskellNTRStack_.sort3N (<=) xs) `seq` return ()),
     ("TRStack.sort3",        \_ xs -> sorted (MergesortHaskellTRStack.sort3 (<=) xs)    `seq` return ()),
     ("TRStack.sortN",        \_ xs -> sorted (MergesortHaskellTRStack.sortN (<=) xs)    `seq` return ()),
     ("TRStack.sort3N",       \_ xs -> sorted (MergesortHaskellTRStack.sort3N (<=) xs)   `seq` return ()),
     ("TRStack_.sort3",       \_ xs -> sorted (MergesortHaskellTRStack_.sort3 (<=) xs)   `seq` return ()),
     ("TRStack_.sortN",       \_ xs -> sorted (MergesortHaskellTRStack_.sortN (<=) xs)   `seq` return ()),
     ("TRStack_.sort3N",      \_ xs -> sorted (MergesortHaskellTRStack_.sort3N (<=) xs)  `seq` return ()),
     ("CBNAcc.sort2",         \_ xs -> sorted (MergesortCoqCbnAcc.sort2 (<=) xs)         `seq` return ()),
     ("CBNAcc.sort3",         \_ xs -> sorted (MergesortCoqCbnAcc.sort3 (<=) xs)         `seq` return ()),
     ("CBNAcc.sortN",         \_ xs -> sorted (MergesortCoqCbnAcc.sortN (<=) xs)         `seq` return ()),
     ("CBVAcc.sort2",         \_ xs -> sorted (MergesortCoqCbvAcc.sort2 (<=) xs)         `seq` return ()),
     ("CBVAcc.sort3",         \_ xs -> sorted (MergesortCoqCbvAcc.sort3 (<=) xs)         `seq` return ()),
     ("CBVAcc.sortN",         \_ xs -> sorted (MergesortCoqCbvAcc.sortN (<=) xs)         `seq` return ())]

File: benchmark/ocaml/benchmark_exp.ml
open Mergesort_ocaml;;

Random.self_init ();;
Gc.set { (Gc.get()) with Gc.minor_heap_size = 2 * 1024 * 1024 * 1024 };;

Benchlib.benchmark "ocaml-random" (List.init 15 (fun i -> Int.shift_left 128 i)) (fun xs -> xs)
  [("List.stable_sort",   StdlibSort.sort                ((<=) : int -> int -> bool));
   ("NTRStack.sort3",     NTRStack.sort3                 ((<=) : int -> int -> bool));
   ("NTRStack.sortN",     NTRStack.sortN                 ((<=) : int -> int -> bool));
   ("NTRStack.sort3N",    NTRStack.sort3N                ((<=) : int -> int -> bool));
   ("NTRStack_.sort3",    NTRStack_.sort3                ((<=) : int -> int -> bool));
   ("NTRStack_.sortN",    NTRStack_.sortN                ((<=) : int -> int -> bool));
   ("NTRStack_.sort3N",   NTRStack_.sort3N               ((<=) : int -> int -> bool));
   ("TRMCStack.sort3",    TRMCStack.sort3                ((<=) : int -> int -> bool));
   ("TRMCStack.sortN",    TRMCStack.sortN                ((<=) : int -> int -> bool));
   ("TRMCStack.sort3N",   TRMCStack.sort3N               ((<=) : int -> int -> bool));
   ("TRMCStack_.sort3",   TRMCStack_.sort3               ((<=) : int -> int -> bool));
   ("TRMCStack_.sortN",   TRMCStack_.sortN               ((<=) : int -> int -> bool));
   ("TRMCStack_.sort3N",  TRMCStack_.sort3N              ((<=) : int -> int -> bool));
   ("TRStack.sort3",      TRStack.sort3                  ((<=) : int -> int -> bool));
   ("TRStack.sortN",      TRStack.sortN                  ((<=) : int -> int -> bool));
   ("TRStack.sort3N",     TRStack.sort3N                 ((<=) : int -> int -> bool));
   ("TRStack_.sort3",     TRStack_.sort3                 ((<=) : int -> int -> bool));
   ("TRStack_.sortN",     TRStack_.sortN                 ((<=) : int -> int -> bool));
   ("TRStack_.sort3N",    TRStack_.sort3N                ((<=) : int -> int -> bool));
   ("CBNAcc.sort2",       Mergesort_coq_cbnacc.sort2     ((<=) : int -> int -> bool));
   ("CBNAcc.sort3",       Mergesort_coq_cbnacc.sort3     ((<=) : int -> int -> bool));
   ("CBNAcc.sortN",       Mergesort_coq_cbnacc.sortN     ((<=) : int -> int -> bool));
   ("CBNAcc.sort2 (TMC)", Mergesort_coq_cbnacc_tmc.sort2 ((<=) : int -> int -> bool));
   ("CBNAcc.sort3 (TMC)", Mergesort_coq_cbnacc_tmc.sort3 ((<=) : int -> int -> bool));
   ("CBNAcc.sortN (TMC)", Mergesort_coq_cbnacc_tmc.sortN ((<=) : int -> int -> bool));
   ("CBVAcc.sort2",       Mergesort_coq_cbvacc.sort2     ((<=) : int -> int -> bool));
   ("CBVAcc.sort3",       Mergesort_coq_cbvacc.sort3     ((<=) : int -> int -> bool));
   ("CBVAcc.sortN",       Mergesort_coq_cbvacc.sortN     ((<=) : int -> int -> bool))]
;;

Benchlib.benchmark "ocaml-smooth" (List.init 15 (fun i -> Int.shift_left 128 i)) (Benchlib.sort_blocks 50)
  [("List.stable_sort",   StdlibSort.sort                ((<=) : int -> int -> bool));
   ("NTRStack.sort3",     NTRStack.sort3                 ((<=) : int -> int -> bool));
   ("NTRStack.sortN",     NTRStack.sortN                 ((<=) : int -> int -> bool));
   ("NTRStack.sort3N",    NTRStack.sort3N                ((<=) : int -> int -> bool));
   ("NTRStack_.sort3",    NTRStack_.sort3                ((<=) : int -> int -> bool));
   ("NTRStack_.sortN",    NTRStack_.sortN                ((<=) : int -> int -> bool));
   ("NTRStack_.sort3N",   NTRStack_.sort3N               ((<=) : int -> int -> bool));
   ("TRMCStack.sort3",    TRMCStack.sort3                ((<=) : int -> int -> bool));
   ("TRMCStack.sortN",    TRMCStack.sortN                ((<=) : int -> int -> bool));
   ("TRMCStack.sort3N",   TRMCStack.sort3N               ((<=) : int -> int -> bool));
   ("TRMCStack_.sort3",   TRMCStack_.sort3               ((<=) : int -> int -> bool));
   ("TRMCStack_.sortN",   TRMCStack_.sortN               ((<=) : int -> int -> bool));
   ("TRMCStack_.sort3N",  TRMCStack_.sort3N              ((<=) : int -> int -> bool));
   ("TRStack.sort3",      TRStack.sort3                  ((<=) : int -> int -> bool));
   ("TRStack.sortN",      TRStack.sortN                  ((<=) : int -> int -> bool));
   ("TRStack.sort3N",     TRStack.sort3N                 ((<=) : int -> int -> bool));
   ("TRStack_.sort3",     TRStack_.sort3                 ((<=) : int -> int -> bool));
   ("TRStack_.sortN",     TRStack_.sortN                 ((<=) : int -> int -> bool));
   ("TRStack_.sort3N",    TRStack_.sort3N                ((<=) : int -> int -> bool));
   ("CBNAcc.sort2",       Mergesort_coq_cbnacc.sort2     ((<=) : int -> int -> bool));
   ("CBNAcc.sort3",       Mergesort_coq_cbnacc.sort3     ((<=) : int -> int -> bool));
   ("CBNAcc.sortN",       Mergesort_coq_cbnacc.sortN     ((<=) : int -> int -> bool));
   ("CBNAcc.sort2 (TMC)", Mergesort_coq_cbnacc_tmc.sort2 ((<=) : int -> int -> bool));
   ("CBNAcc.sort3 (TMC)", Mergesort_coq_cbnacc_tmc.sort3 ((<=) : int -> int -> bool));
   ("CBNAcc.sortN (TMC)", Mergesort_coq_cbnacc_tmc.sortN ((<=) : int -> int -> bool));
   ("CBVAcc.sort2",       Mergesort_coq_cbvacc.sort2     ((<=) : int -> int -> bool));
   ("CBVAcc.sort3",       Mergesort_coq_cbvacc.sort3     ((<=) : int -> int -> bool));
   ("CBVAcc.sortN",       Mergesort_coq_cbvacc.sortN     ((<=) : int -> int -> bool))]
;;
